src/CartoonGAN/index.js

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2018 ml5
+//
+// This software is released under the MIT License.
+// https://opensource.org/licenses/MIT
+
+/* eslint prefer-destructuring: ["error", {AssignmentExpression: {array: false}}] */
+/* eslint no-await-in-loop: "off" */
+/*
+* CartoonGAN: see details about the [paper](http://openaccess.thecvf.com/content_cvpr_2018/papers/Chen_CartoonGAN_Generative_Adversarial_CVPR_2018_paper.pdf)
+*/
+
+import * as tf from '@tensorflow/tfjs';
+import callCallback from '../utils/callcallback';
+import p5Utils from '../utils/p5Utils';
+
+const modelPath = {
+    'hosoda': 'https://raw.githubusercontent.com/leemengtaiwan/tfjs-models/master/cartoongan/tfjs_json_models/hosoda/model.json',
+    'miyazaki': 'https://raw.githubusercontent.com/Derek-Wds/training_CartoonGAN/master/tfModels/Miyazaki/model.json'
+};
+
+class Cartoon {
+    /**
+     * Create a CartoonGan model.
+     * @param {String} modelIdentifier - Required. The name of pre-inluded model or the url path to your model.
+     * @param {function} callback - Required. A function to run once the model has been loaded.
+     */
+    constructor(modelIdentifier, callback) {
+        this.modelUrl = modelPath[modelIdentifier] ? modelPath[modelIdentifier] : modelIdentifier;
+        this.ready = false;
+        this.model = {};
+        this.ready = callCallback(this.loadModel(this.modelUrl), callback);
+    }
+
+    /* load tfjs model that is converted by tensorflowjs with graph and weights */
+    async loadModel(modelUrl) {
+        this.model = await tf.loadGraphModel(modelUrl);
+        return this;
+    }
+
+
+    // todo: add p5 image support as input
+    /**
+     * generate an img based on input Image.
+     * @param {HTMLImageElement | HTMLCanvasElement} src the source img you want to transfer.
+     * @param {function} callback
+     */
+    async generate(src, callback) {
+        if( !(src instanceof HTMLImageElement || src instanceof HTMLCanvasElement) ){
+            throw new Error (`Invalid input type: ${typeof(src)}\nExpected HTMLImgElement, p5Image or HTMLCanvasElement`);
+        }
+        await this.ready;
+        return callCallback(this.generateInternal(src), callback);
+    }
+
+    async generateInternal(src) {
+        await this.ready;
+        let img = tf.browser.fromPixels(src);
+        if (img.shape[0] !== 256 || img.shape[1] !== 256) {
```

Comment:
This can be innefective:

```JS
const {shape: [width, height, colorChannelNumber ] = [] } = img;
const hasImageTheRightShape = width === 256 & height === 256 && colorChannelNumber === 3;
if(!hasImageTheRightShape) {
throw new Error(`Image should have 256*256 with 3 color channel. But ${width}x${height} with ${colorChannelNumber} color channels found!`);
}
...
```